File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.py
import heapq
def solution(operations):
    answer = [0,0]
    h = []
    for i in range(len(operations)):
        if operations[i] == 'D 1':
            if len(h) > 0:
                h.pop()
        elif operations[i] == 'D -1':
            if len(h) > 0:
                heapq.heappop(h)
        else:
            a, b = operations[i].split()
            heapq.heappush(h, int(b))
    
    if len(h) > 0:
        answer = [max(h), min(h)]
    return answer 

File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.java
import java.util.*;

class Solution {
    public int[] solution(String[] operations) {
        PriorityQueue<Integer> max = new PriorityQueue<>(Collections.reverseOrder());
        PriorityQueue<Integer> min = new PriorityQueue<>();
        
        for(String op : operations){
            String[] o = op.split(" ");
            if(o[0].equals("I")){
                max.offer(Integer.parseInt(o[1]));
                min.offer(Integer.parseInt(o[1]));
            }else if(o[1].equals("1") && max.size() > 0){
                int pop = max.poll();
                min.remove(pop);
            }else if(o[1].equals("-1") && max.size() > 0) {
                int pop = min.poll();
                max.remove(pop);
            }
        }
        
        int[] answer = new int[] {0, 0};
        answer[0] = max.size() > 0 ? max.peek() : 0;
        answer[1] = min.size() > 0 ? min.peek() : 0;
        return answer;
    } 
} 